### PR TITLE
Add missing generic constraints to generated methods

### PR DESCRIPTION
**Pull request by trejjam**
# Add missing generic constraints to generated methods

## Description

A method `partial TTarget Map<TSource, TTarget>(TSource source) where TSource : class, TTarget : class;` is now generated as
`partial TTarget Map<TSource, TTarget>(TSource source) { ... }`, which is invalid code, generic constraints need to be specified in partial implementation as well.

## Checklist

- [x] The existing code style is followed
- [x] The commit message follows our guidelines
- [x] Performed a self-review of my code
- [x] Hard-to-understand areas of my code are commented
- [x] The documentation is updated (as applicable)
- [x] Unit tests are added/updated
- [x] Integration tests are added/updated (as applicable, especially if feature/bug depends on roslyn or framework version in use)

### CODE REVIEW

File: src/Riok.Mapperly/Descriptors/Mappings/UserMappings/UserDefinedNewInstanceGenericTypeMapping.cs [hunk 1]
```diff
@@ -35,7 +37,9 @@ ITypeSymbol objectType
     public override MethodDeclarationSyntax BuildMethod(SourceEmitterContext ctx)
     {
         var methodSyntax = (MethodDeclarationSyntax)Method.DeclaringSyntaxReferences.First().GetSyntax();
-        return base.BuildMethod(ctx).WithTypeParameterList(methodSyntax.TypeParameterList);
+        return base.BuildMethod(ctx)
+            .WithTypeParameterList(methodSyntax.TypeParameterList)
+            .WithConstraintClauses(List(GetTypeParameterConstraintClauses()));
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/Mappings/UserMappings/UserDefinedNewInstanceGenericTypeMapping.cs:42]: Is there a reason for not simply using `methodSyntax.ConstraintClauses` here?

File: test/Riok.Mapperly.Tests/GeneratedMethod.cs [hunk 2]
```diff
@@ -8,15 +9,26 @@ public GeneratedMethod(MethodDeclarationSyntax declarationSyntax)
...
+    public string? ConstraintClauses { get; }
+
     public string Body { get; }
 
+    private static string? ExtractParameterConstraints(SyntaxList<TypeParameterConstraintClauseSyntax> typeParameterConstraints)
+    {
+        if (typeParameterConstraints.Count == 0)
+            return null;
+
+        return typeParameterConstraints.ToFullString().Trim(' ', '\r', '\n').Trim('\r', '\n').ReplaceLineEndings();
```
Comment (latonz) [inline test/Riok.Mapperly.Tests/GeneratedMethod.cs:29]: Is the second trim really needed?

File: test/Riok.Mapperly.Tests/GeneratedMethod.cs [hunk 3]
```diff
@@ -8,15 +9,26 @@ public GeneratedMethod(MethodDeclarationSyntax declarationSyntax)
...
         Signature = $"{declarationSyntax.ReturnType.ToString()} {Name}{declarationSyntax.ParameterList.ToString().Trim()}";
+        ConstraintClauses = ExtractParameterConstraints(declarationSyntax.ConstraintClauses);
         Body = ExtractBody(declarationSyntax);
     }
 
     public string Name { get; }
 
     public string Signature { get; }
 
+    public string? ConstraintClauses { get; }
```
Comment (latonz) [inline test/Riok.Mapperly.Tests/GeneratedMethod.cs:20]: Couldn't we simply add this to the signature? IMO this is part of the signature in C#...

**Review by latonz — CHANGES_REQUESTED** [review]
Thanks for the fix, embarrassing that I didn't spot this earlier.

File: src/Riok.Mapperly/Descriptors/Mappings/UserMappings/UserDefinedNewInstanceGenericTypeMapping.cs (hunk 1 above)
Comment (trejjam) [reply src/Riok.Mapperly/Descriptors/Mappings/UserMappings/UserDefinedNewInstanceGenericTypeMapping.cs:42]: Yes, it modifies types to by `FullyQualifiedIdentifier`:
image

File: test/Riok.Mapperly.Tests/GeneratedMethod.cs (hunk 2 above)
Comment (trejjam) [reply test/Riok.Mapperly.Tests/GeneratedMethod.cs]: Not at all

File: test/Riok.Mapperly.Tests/GeneratedMethod.cs (hunk 3 above)
Comment (trejjam) [reply test/Riok.Mapperly.Tests/GeneratedMethod.cs:20]: We can (I was searching for it there originally). I personally like that it's also a separate property (for these constraints-specific checks).

Is there a problem if it is duplicated i.e.:
```cs
ConstraintClauses = ExtractParameterConstraints(declarationSyntax.ConstraintClauses);
Signature = $"{declarationSyntax.ReturnType.ToString()} {Name}{declarationSyntax.ParameterList.ToString().Trim()}{ConstraintClauses}";
```

Comment (latonz) [reply test/Riok.Mapperly.Tests/GeneratedMethod.cs:20]: Works with me 👍

**Review by latonz — APPROVED** [review]

**Pull request merged**